### PR TITLE
New Rule S6967: ModelState.IsValid should be called in controller actions

### LINKED ISSUE #9095: New Rule S6967: ModelState.IsValid should be called in controller actions
RSPEC: https://github.com/SonarSource/rspec/pull/3856

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #9095

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    private static void ProcessControllerMethods(SonarSyntaxNodeReportingContext context)
+    {
+        var method = (MethodDeclarationSyntax)context.Node;
+        if (context.SemanticModel.GetDeclaredSymbol(context.Node) is IMethodSymbol methodSymbol
+            && methodSymbol.IsControllerMethod()
+            && !HasActionFilterAttribute(methodSymbol)
+            && methodSymbol.Parameters.Any(x => HasValidationAttribute(x) || IsValidatable(x.Type)))
+        {
+            var walker = new ModelStateFinder(context.SemanticModel);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:62]: As an alternative to the walker, you could use `RegisterCodeBlockStartAction` and in there `RegisterNodeAction` instead.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            walker.Visit(method.Body);
+            if (!walker.ValidatesModel)
+            {
+                context.ReportIssue(Rule, method.Identifier);
+            }
+        }
+    }
+
+    private static bool IsValidatable(ITypeSymbol type) => type != null
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:71]: Why do we need to check this? I think every model is validated no matter if it has any validation attributes/interfaces. E.g.:

A model like this
```cs
public class Model
{
    public int Id { get; set; }
}
```
with input like this:
```json
{
  "Id": "Boom"
}
```
should set the ModelState.IsValid to false (Needs confirmation).
See e.g. here https://stackoverflow.com/a/36894368 (might be related to ASP.NET 4.x, though)

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/CallModelStateIsValid.cs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+    [AcceptVerbs("GET", "POST", "PUT", "DELETE", "PATCH")]
+    [Route("/[controller]/many")]
+    public string Many([Required, FromQuery, EmailAddress] string email)        // Noncompliant
+    {
+        return "Hello!";
+    }
+
+    [HttpPost("/[controller]/without-validation")]
+    public string DtoWithoutValidation(DtoWithoutValidation dto)                // Compliant, DtoWithoutValidation does not have any validation attributes.
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/CallModelStateIsValid.cs:136]: You could still post invalid json, which should set ModelState.IsValid to false.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+               && semanticModel.GetSymbolInfo(node.Expression).Symbol is IMethodSymbol method
+               && method.ContainingType.Is(KnownType.Microsoft_AspNetCore_Mvc_ControllerBase))
+            {
+                ValidatesModel = true;
+            }
+        }
+
+        public override void VisitMemberAccessExpression(MemberAccessExpressionSyntax node)
+        {
+            if (node.GetName() == "IsValid"
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:109]: Also `ValidationState` https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.modelbinding.modelstatedictionary.validationstate?view=aspnetcore-8.0

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/CallModelStateIsValid.cs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        return "Hello!";
+    }
+}
+
+public class CompliantController: ControllerBase
+{
+    [HttpGet("/[controller]")]
+    public string Get([Required, FromQuery] string id)                           // Compliant
+    {
+        if (!ModelState.IsValid)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/CallModelStateIsValid.cs:41]: Maybe a parameterized UT. Different access:
```cs
ModelState?.IsValid
ModelState!.IsValid
ModelState!?.IsValid
ModelState?.IsValid!
ModelState is { IsValid: false }
this is { ModelState.IsValid: false }
this is { ModelState: { IsValid: false } }
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                base.Visit(node);
+            }
+        }
+
+        public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+        {
+            if (node.GetName() == "TryValidateModel"
+               && semanticModel.GetSymbolInfo(node.Expression).Symbol is IMethodSymbol method
+               && method.ContainingType.Is(KnownType.Microsoft_AspNetCore_Mvc_ControllerBase))
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:101]: [TryValidateModel](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.tryvalidatemodel?view=aspnetcore-8.0) is virtual. Add a test case where `TryValidateModel` is overridden.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStart =>
+        {
+            if (compilationStart.Compilation.ReferencesControllers()
+                && !compilationStart.Compilation.Assembly.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ApiControllerAttribute))
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:39]: It's not checking whether it's available, but rather if the attribute was applied on the assembly. [See this test case](https://github.com/SonarSource/sonar-dotnet/blob/6d58f9ba27bfde28b9f1cf7e193c65c8d3445342/analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/CallModelStateIsValidTest.cs#L49).

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                compilationStart.RegisterSymbolStartAction(symbolStart =>
+                {
+                    var type = (INamedTypeSymbol)symbolStart.Symbol;
+                    if (type.DerivesFrom(KnownType.Microsoft_AspNetCore_Mvc_ControllerBase)
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:44]: The `ModelState` property comes from the `ControllerBase` class, so the rule only cares about classes that derive from that.
A regular class that's decorated with the `[Controller]` or `[ApiController]` attribute doesn't have that property, so the rule can't ask the user to perform validation with it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs (hunk 2 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:71]: That's a good point, and it makes the rule so much easier! I'll update the RSPEC.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs]: Did you validate/confirm this behavior in a real Asp.Net application? Is `ModelState.IsValid` set or is the pipeline returning an `500 Bad request` before the method is even called?

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Some more test cases for GetIdentifier and ConditionalAccess are needed and some minor nitpicks.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 9]
```diff
@@ -172,6 +172,7 @@ node switch
                 AttributeArgumentSyntax { NameEquals.Name.Identifier: var identifier } => identifier,
                 AttributeSyntax { Name: { } name } => GetIdentifier(name),
                 BaseTypeDeclarationSyntax { Identifier: var identifier } => identifier,
+                ConditionalAccessExpressionSyntax { WhenNotNull: var rightSide } => GetIdentifier(rightSide),
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:185]: It seems to work as intended. I added more tests.

File: analyzers/tests/SonarAnalyzer.Test/Extensions/SyntaxNodeExtensionsTest.cs [hunk 10]
```diff
@@ -1094,6 +1094,12 @@ public class C
...
+        [DataRow("""void M(string s) { var length = $$s?.Length$$; }""", "Length")]                                     // ConditionalAccessExpressionSyntax
+        [DataRow("""void M(string s) { var length = $$s?.ToLower()?.ToUpper()$$; }""", "ToUpper")]                      // ConditionalAccessExpressionSyntax
+        [DataRow("""void M(string s) { var length = $$s.ToLower()?.ToUpper()$$; }""", "ToUpper")]                       // ConditionalAccessExpressionSyntax + MemberAccessExpressionSyntax
+        [DataRow("""void M(string s) { var length = $$s?.ToLower().ToUpper()$$; }""", "ToUpper")]                       // ConditionalAccessExpressionSyntax + MemberAccessExpressionSyntax
+        [DataRow("""void M(string s) { var length = $$s?.ToLower().ToUpper()?.ToLower().ToUpper()$$; }""", "ToUpper")]  // ConditionalAccessExpressionSyntax + MemberAccessExpressionSyntax
+        [DataRow("""void M(string s) { var length = $$s.ToLower().ToUpper().ToLower().ToUpper()$$; }""", "ToUpper")]    // MemberAccessExpressionSyntax
         [DataRow("""$$Test() { }$$""", "Test")]                                       // ConstructorDeclarationSyntax
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Extensions/SyntaxNodeExtensionsTest.cs:1103]:
```suggestion
        [DataRow("""void M(string s) { var length = $$s.ToLower().ToUpper()$$.Format("").Normalize(); }""", "ToUpper")]
        [DataRow("""void M(string s) { var length = $$s.ToLower().ToUpper()$$?.Format("").Normalize(); }""", "ToUpper")]
        [DataRow("""void M(string s) { var length = s.ToLower().?$$ToUpper()?.Format("").Normalize()$$; }""", "Normalize")]
        [DataRow("""$$Test() { }$$""", "Test")]                                       // ConstructorDeclarationSyntax
```

File: analyzers/tests/SonarAnalyzer.Test/Helpers/AspNetMvcHelperTest.cs [hunk 11]
```diff
@@ -37,6 +37,10 @@ public class Foo : System.Web.Mvc.Controller
     protected void ProtectedFoo() { }
     internal void InternalFoo() { }
     private void PrivateFoo() { }
+    public static void StaticFoo() { }
+    public void OutFoo(out string arg) { arg = null; }
+    public void RefFoo(ref string arg) { }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Helpers/AspNetMvcHelperTest.cs:42]: NIT: Add the other ref kinds as well under test:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/method-parameters#reference-parameters

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Helpers/AspNetMvcHelperTest.cs:42]: These are all ordinary methods. We should also put at least one other [method kind](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.methodkind?view=roslyn-dotnet-4.7.0) under test. I think a constructor with parameters should be enough to cover the "Ordinary" check.

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Helpers/AspNetMvcHelperTest.cs:42]: Some [method modifiers](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1561-general) are also interesting to put under test:
* new, virtual, override, sealed -> okay
* abstract -> okay
* extern -> Not okay (extern method need to be static anyway)
* unsafe -> I don't know. It shouldn't matter, so probably okay.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/CallModelStateIsValidTest.cs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+                public MovieController(IValidator<Movie> validator)
+                {
+                    _validator = validator;
+                }
+
+                [HttpPost("/[controller]")]
+                public string Add(Movie movie)  // Compliant - uses FluentValidation
+                {
+                    if (!_validator.Validate(movie).IsValid)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/CallModelStateIsValidTest.cs:100]: Nice. (This pattern would probably also work for POCO controllers).

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/CallModelStateIsValid.cs [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+    {
+        return "Hello!";
+    }
+}
+
+[Controller]
+public class ControllerThatDoesNotInheritFromControllerBase
+{
+    [HttpPost("/[controller]")]
+    public string Add(Movie movie)                                              // Compliant, ModelState is not available in this context.
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/CallModelStateIsValid.cs:109]: You may want to add a FN example with a FluentValidation dependency, where the validation is missing.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private static bool HasActionFilterAttribute(ISymbol symbol) =>
+        symbol.GetAttributes().Any(x => x.AttributeClass.DerivesFrom(KnownType.Microsoft_AspNetCore_Mvc_Filters_ActionFilterAttribute));
+
+    private static bool IsCheckingValidityProperty(SyntaxNode node, SemanticModel semanticModel)
+    {
+        var nodeIdentifier = node.GetIdentifier();
+        return nodeIdentifier?.ValueText is "IsValid" or "ValidationState"
+            && semanticModel.GetSymbolInfo(nodeIdentifier.Value.Parent).Symbol is IPropertySymbol propertySymbol
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs:94]:
```suggestion
        return node.GetIdentifier() is { } nodeIdentifier
            && nodeIdentifier.ValueText is "IsValid" or "ValidationState"
            && semanticModel.GetSymbolInfo(nodeIdentifier.Parent).Symbol is IPropertySymbol propertySymbol
```

File: analyzers/tests/SonarAnalyzer.Test/Extensions/SyntaxNodeExtensionsTest.cs (hunk 10 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Extensions/SyntaxNodeExtensionsTest.cs:1103]: These cases seem to be missing

File: analyzers/tests/SonarAnalyzer.Test/Helpers/AspNetMvcHelperTest.cs (hunk 11 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Helpers/AspNetMvcHelperTest.cs:45]: The constructor test is missing. https://github.com/SonarSource/sonar-dotnet/pull/9183/files#r1579066058

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/CallModelStateIsValidTest.cs (hunk 12 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/CallModelStateIsValidTest.cs:100]: I was referring to injecting the `IValidator` service. That way, you can validate in a POCO controller action. Please also take a look at this comment https://github.com/SonarSource/sonar-dotnet/pull/9183/files#r1579080060 which you fixed.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Some test cases are still missing, otherwise looks good.

**Review by martin-strecker-sonarsource — APPROVED** [review]

**Comment by martin-strecker-sonarsource** [comment]
LGTM

**Pull request merged**